src/com/rcv/TieBreak.java

```diff
@@ -87,41 +93,61 @@ String nonLosingCandidateDescription() {
     return nonselected;
   }
 
-  String loser() {
-    return loser;
-  }
-
   String explanation() {
     return explanation;
   }
 
   // function: breakTie
   // purpose: execute the tiebreak logic given the tiebreak rule in use
   // returns: losing candidate
-  private String breakTie() {
-    // return value
-    String losingCandidate;
+  String selectLoser() {
     switch (tieBreakMode) {
       case INTERACTIVE:
-        losingCandidate = doInteractive();
+        loser = doInteractive();
         break;
       case RANDOM:
-        losingCandidate = doRandom();
+        loser = doRandom();
+        break;
+      case GENERATE_PERMUTATION:
+      case USE_PERMUTATION_IN_CONFIG:
+        loser = doPermutationSelection();
         break;
       default:
         // handle tiebreaks which involve previous round tallies
         // loser: will be set if there is a previous round count loser
         // it will be null if candidates were still tied at first round
-        String loser = doPreviousRounds();
-        if (loser != null) {
-          losingCandidate = loser;
+        String previousRoundsLoser = doPreviousRounds();
+        if (previousRoundsLoser != null) {
+          loser = previousRoundsLoser;
         } else if (tieBreakMode == Tabulator.TieBreakMode.PREVIOUS_ROUND_COUNTS_THEN_INTERACTIVE) {
-          losingCandidate = doInteractive();
+          loser = doInteractive();
         } else {
-          losingCandidate = doRandom();
+          loser = doRandom();
         }
     }
-    return losingCandidate;
+    return loser;
+  }
+
+  // function: doPermutationSelection
+  // purpose: select the loser based on the provided permutation order
+  // returns: selected loser
+  private String doPermutationSelection() {
+    // loser to return
+    String selectedCandidate = null;
+    // create a set to simplify matching logic
+    Set<String> tiedCandidatesSet = new HashSet<>(tiedCandidates);
+
+    // start from end of list and look for a matching candidate
+    for (int i = candidatePermutation.size() - 1; i >= 0; i--) {
+      // the current candidate we're considering from the permutation
+      String candidate = candidatePermutation.get(i);
+      if (tiedCandidatesSet.contains(candidate)) {
+        selectedCandidate = candidate;
+        break;
+      }
+    }
+    explanation = "The losing candidate appeared last in the tie-breaking permutation list.";
```

Comment:
losing candidate appeared "later"

Reply:
This looks solid, I only wonder what happens if UWI is involved in the tiebreak?

Reply:
I was thinking we could add optional "tieBreakOrder" field into the Candidate object of RawElectionConfig.  If user specifies USE_PERMUTATION_IN_CONFIG then validation code can make sure user ordering is valid before proceeding.

Reply:
> losing candidate appeared "later"

I changed it to "latest." (Remember that there can be more than 2 people tied.)

> This looks solid, I only wonder what happens if UWI is involved in the tiebreak?

The tabulator drops UWIs before it does anything else, so they can't be involved in a tie-break scenario.

> I was thinking we could add optional "tieBreakOrder" field into the Candidate object of RawElectionConfig. If user specifies USE_PERMUTATION_IN_CONFIG then validation code can make sure user ordering is valid before proceeding.

I thought about that, but decided to start with this approach (just use the actual ordering in the file) because it's simpler. I'm fine with changing to do it the way you suggested later if the group agrees that it's better.

Reply:
Yeah, your approach is simpler so that's fine.